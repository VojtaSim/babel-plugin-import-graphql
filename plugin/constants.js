export const newlinePattern = /(\r\n|\r|\n)+/

export const importPattern = /^# *?import +?(?:\* +?from +?)?(?:'|"|`)(.+?)(?:'|"|`)(?:\r\n|\r|\n)*$/gm
