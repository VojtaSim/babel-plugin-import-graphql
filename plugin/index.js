import path, { dirname, delimiter } from 'path'
import { existsSync, readFileSync } from 'fs'
import { transformSync } from '@babel/core'
import gql from 'graphql-tag'

import { createDocPerOp } from './multi-op'

const newlinePattern = /(\r\n|\r|\n)+/
let resolve
export default ({ types: t }) => ({
  manipulateOptions({ resolveModuleSource }) {
    if (!resolve) {
      resolve = resolveModuleSource || ((src, file) => path.resolve(dirname(file), src))
    }
  },
  visitor: {
    ImportDeclaration: {
      exit(curPath, { file: { opts: { filename: babelPath } }, opts: { nodePath } }) {
        const importNames = curPath.node.specifiers
        const importPath = curPath.node.source.value
        if (importPath.endsWith('.graphql') || importPath.endsWith('.gql')) {
          // Find and fetch the file, using node resolution/NODE_PATH if necessary.
          const fallbackPaths = nodePath ? nodePath.split(delimiter) : process.env.NODE_PATH
          let absPath = resolve(importPath, babelPath)
          if (!existsSync(absPath)) absPath = require.resolve(importPath, { paths: fallbackPaths })
          const source = readFileSync(absPath).toString()

          // If the file doesn't contain query/mutation/subscription, insert raw text, else parse.
          if (isSchemaLike(source)) {
            curPath.replaceWith(buildVariableAST(source, importNames[0].local.name))
          } else {
            const doc = processDoc(createDoc(source, absPath))

            const replacements = buildReplacements(createDocPerOp(doc), importNames)
            replacements.length > 1
              ? curPath.replaceWithMultiple(replacements)
              : curPath.replaceWith(replacements[0])
          }
        }

        function isSchemaLike(source) {
          const content = source
            .split(newlinePattern)
            .filter(line => !newlinePattern.test(line))
            .filter(line => !line.startsWith('#'))
            .filter(line => line.length > 0)

          const operationsPattern = /^(fragment|query|mutation|subscription)/
          return !operationsPattern.test(content[0])
        }

        function buildReplacements(docs, specifiers) {
          return specifiers.map(({ type, imported, local }) => {
            switch (type) {
              case 'ImportDefaultSpecifier':
                return buildVariableAST(docs.default, local.name)
              case 'ImportSpecifier':
                return buildVariableAST(docs[imported.name] || docs.default, local.name)
              case 'ImportNamespaceSpecifier':
                return buildVariableAST(docs, local.name)
              default:
                throw new Error(`Unexpected import specifier type: ${type}`)
            }
          })
        }

        function buildVariableAST(graphqlAST, importName) {
          const { ast } = transformSync(`var ${importName} = ${JSON.stringify(graphqlAST)}`, {
            ast: true
          })
          return ast.program.body[0]
        }
      }
    }
  }
})

function createDoc(source, absPath) {
  let ast = null
  let fragmentDefs = []

  return {
    processFragments() {
      processImports(getImportStatements(source), absPath)

      function getImportStatements(src) {
        return src.split(newlinePattern).filter(line => line.startsWith('#import'))
      }

      function processImports(imports, relFile) {
        imports.forEach(statement => {
          const fragmentPath = statement.split(/[\s\n]+/g)[1].slice(1, -1)
          const absFragmentPath = resolve(fragmentPath, relFile)
          const fragmentSource = readFileSync(absFragmentPath.replace(/'/g, '')).toString()
          const subFragments = getImportStatements(fragmentSource)
          if (subFragments.length > 0) processImports(subFragments, absFragmentPath)
          // prettier-ignore
          fragmentDefs = [...gql`${fragmentSource}`.definitions, ...fragmentDefs]
        })
      }
    },
    parse() {
      // prettier-ignore
      const parsedAST = gql`${source}`
      parsedAST.definitions = [...parsedAST.definitions, ...fragmentDefs]
      ast = parsedAST
    },
    dedupeFragments() {
      let seenNames = {}
      ast.definitions = ast.definitions.filter(def => {
        if (def.kind !== 'FragmentDefinition') return true
        return seenNames[def.name.value] ? false : (seenNames[def.name.value] = true)
      })
    },
    makeSourceEnumerable() {
      const newAST = JSON.parse(JSON.stringify(ast))
      newAST.loc.source = ast.loc.source
      ast = newAST
    },
    get ast() {
      return ast
    }
  }
}

function processDoc(doc) {
  doc.processFragments()
  doc.parse()
  doc.dedupeFragments()
  doc.makeSourceEnumerable()
  return doc.ast
}
