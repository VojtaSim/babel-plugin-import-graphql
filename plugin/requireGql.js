import { readFileSync } from 'fs'
import path, { isAbsolute, join, dirname } from 'path'
import gql from 'graphql-tag'

import { createDocPerOp } from './multi-op'

const newlinePattern = /(\r\n|\r|\n)+/

export const defaultResolve = (src, file) => path.resolve(dirname(file), src)

export const requireGql = (filepath, { resolve = defaultResolve, nowrap = true } = {}) => {
  filepath = isAbsolute(filepath) ? filepath : join(callerDirname(), filepath)
  const source = readFileSync(filepath).toString()

  // If the file doesn't contain ops return raw text, else parse  and return docs object.
  if (isSchemaLike(source)) return source

  const docs = createDocPerOp(processDoc(createDoc(source, filepath, resolve)))
  return nowrap && Object.keys(docs).length <= 2 ? docs.default : docs
}

function callerDirname() {
  // To avoid dependencies, I borrowed this from gh (sindresorhus/callsites/blob/master/index.js)
  const _ = Error.prepareStackTrace
  Error.prepareStackTrace = (_, stack) => stack
  const stack = new Error().stack.slice(1)
  Error.prepareStackTrace = _
  // End borrowed code

  const caller = stack.find(c => c.getTypeName() !== null)
  return dirname(caller.getFileName())
}

function isSchemaLike(source) {
  const content = source
    .split(newlinePattern)
    .filter(line => !newlinePattern.test(line))
    .filter(line => !line.startsWith('#'))
    .filter(line => line.length > 0)
    .map(line => line.trimLeft())

  const operationsPattern = /^(fragment|query|mutation|subscription)/
  return !operationsPattern.test(content[0])
}

function createDoc(source, filepath, resolve) {
  let ast = null
  let fragmentDefs = []

  return {
    processFragments() {
      processImports(getImportStatements(source), filepath)

      function getImportStatements(src) {
        return src.split(newlinePattern).filter(line => line.startsWith('#import'))
      }

      function processImports(imports, relFile) {
        imports.forEach(statement => {
          const fragmentPath = statement.split(/[\s\n]+/g)[1].slice(1, -1)
          const absFragmentPath = resolve(fragmentPath, relFile)
          const fragmentSource = readFileSync(absFragmentPath.replace(/'/g, '')).toString()
          const subFragments = getImportStatements(fragmentSource)
          if (subFragments.length > 0) processImports(subFragments, absFragmentPath)
          // prettier-ignore
          fragmentDefs = [...gql`${fragmentSource}`.definitions, ...fragmentDefs]
        })
      }
    },
    parse() {
      // prettier-ignore
      const parsedAST = gql`${source}`
      parsedAST.definitions = [...parsedAST.definitions, ...fragmentDefs]
      ast = parsedAST
    },
    dedupeFragments() {
      let seenNames = {}
      ast.definitions = ast.definitions.filter(def => {
        if (def.kind !== 'FragmentDefinition') return true
        return seenNames[def.name.value] ? false : (seenNames[def.name.value] = true)
      })
    },
    makeSourceEnumerable() {
      const newAST = JSON.parse(JSON.stringify(ast))
      newAST.loc.source = ast.loc.source
      ast = newAST
    },
    get ast() {
      return ast
    }
  }
}

function processDoc(doc) {
  doc.processFragments()
  doc.parse()
  doc.dedupeFragments()
  doc.makeSourceEnumerable()
  return doc.ast
}
