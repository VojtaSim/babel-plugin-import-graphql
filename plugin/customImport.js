import { readFileSync } from 'fs'

import { newlinePattern } from './constants'

export function getFilepaths(src, relFile, resolve) {
  const imports = src.split(newlinePattern).filter(line => line.startsWith('#import'))
  return imports.map(statement => {
    const importPath = statement.split(/[\s\n]+/g)[1].slice(1, -1)
    return resolve(importPath, relFile)
  })
}

export function getSources(filepath, acc = []) {
  const importSrc = readFileSync(filepath.replace(/'/g, '')).toString()
  const nestedPaths = getFilepaths(importSrc, filepath)
  const srcs = nestedPaths.length > 0 ? nestedPaths.map(fp => getSources(fp, acc)) : [importSrc]
  return [...acc, ...srcs]
}
