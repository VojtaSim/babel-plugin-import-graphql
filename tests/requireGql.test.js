import { join } from 'path'

import { requireGql } from '../plugin/requireGql'

describe('parse client-side graphql file at runtime', () => {
  test('absolute filepath argument', () => {
    const result = requireGql(join(__dirname, './fixtures/shared/named.graphql'))
    expect(typeof result).toBe('object')
  })

  test('relative filepath argument', () => {
    const result = requireGql('./fixtures/shared/named.graphql')
    expect(typeof result).toBe('object')
  })

  test('single named query', () => {
    const result = requireGql('./fixtures/shared/named.graphql')
    expect(result.kind).toBe('Document')
  })

  test('single named query with { nowrap: false }', () => {
    const { named } = requireGql('./fixtures/shared/named.graphql', { nowrap: false })
    expect(named.kind).toBe('Document')
  })

  test('single unnamed query', () => {
    const result = requireGql('./fixtures/shared/unnamed.graphql')
    expect(result.kind).toBe('Document')
  })

  test('multiple operations', () => {
    const result = requireGql('./fixtures/shared/multiple.graphql')
    const { first, second, third } = result
    expect(first).toEqual(result.default)
    expect(first.kind).toBe('Document')
    expect(second.kind).toBe('Document')
    expect(third.kind).toBe('Document')
  })

  test('query with nested fragment', () => {
    const full = requireGql('./fixtures/requireGql/fragment/nested/partial.graphql')
    expect(full.kind).toBe('Document')
    expect(full.definitions).toHaveLength(2)
    expect(full.definitions[0].kind).toBe('OperationDefinition')
    expect(full.definitions[1].kind).toBe('FragmentDefinition')
  })
})

describe('parse schema file at runtime', () => {
  test('returns raw source text', () => {
    const schema = requireGql('./fixtures/shared/schema.graphql')
    expect(schema).toMatchSnapshot()
  })
})
