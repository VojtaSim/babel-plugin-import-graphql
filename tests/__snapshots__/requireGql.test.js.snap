// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse schema file at runtime returns raw source text 1`] = `
"schema {
  query: Query
}

type Query {
  test: String
}
"
`;
